compiler/rustc_middle/src/mir/interpret/error.rs

```diff
@@ -435,8 +435,10 @@ impl<T: Any> AsAny for T {
 }
 
 /// A trait for machine-specific errors (or other "machine stop" conditions).
-pub trait MachineStopType: AsAny + fmt::Display + Send {}
-impl MachineStopType for String {}
+pub trait MachineStopType: AsAny + fmt::Display + Send {
+    /// If `true`, emit a hard error instead of going through the `CONST_ERR` lint
+    fn is_hard_err(&self) -> bool;
```

Comment:
Maybe add a default implementation (returning `false`, presumably) so that we don't need to adjust this in Miri?